sysl2/sysl/validate.go

```diff
@@ -0,0 +1,355 @@
+					Message: fmt.Sprintf("In view '%s', type '%s' is missing", viewName, gk),
+					MsgType: ERROR,
+				})
+			}
+		} else {
+			delete(implAttrNames, gk)
+		}
+	}
+
+	for attrName := range implAttrNames {
+		validationMsgs = append(validationMsgs, ValidationMsg{
+			Message: fmt.Sprintf("In view '%s', excess attribute is defined: '%s'", viewName, attrName),
+			MsgType: ERROR,
+		})
+		delete(implAttrNames, attrName)
+	}
+
+	return validationMsgs
+}
+
+func compareOneOf(grammarSpec, implAttrs map[string]*sysl.Type,
+	implAttrNames map[string]struct{}, viewName, typeName string) []ValidationMsg {
+	validationMsgs := make([]ValidationMsg, 0, 10)
+	matching := true
+
+	for _, one := range grammarSpec[typeName].GetOneOf().GetType() {
+		name := one.GetTypeRef().GetRef().GetPath()[0]
+
+		if strings.Index(name, "__Choice_Combination_") == 0 {
+			if len(implAttrs) == 1 {
+				continue
+			}
+
+			validationMsgs = append(validationMsgs, compareSpecVsImpl(grammarSpec,
+				grammarSpec[name].GetTuple().GetAttrDefs(), implAttrs, implAttrNames, viewName)...)
+			break
+		} else {
+			if _, exists := implAttrs[name]; !exists {
+				matching = false
+			} else {
+				matching = true
+				delete(implAttrNames, name)
+				break
+			}
+		}
+	}
+
+	if !matching {
+		implAttrName := ""
+		for k := range implAttrs {
+			implAttrName = k
+		}
+		validationMsgs = append(validationMsgs, ValidationMsg{
+			Message: fmt.Sprintf("In view '%s', invalid choice has been made as : '%s'", viewName, implAttrName),
+			MsgType: ERROR})
+	}
+
+	return validationMsgs
+}
+
+func hasSameType(type1 *sysl.Type, type2 *sysl.Type) (bool, string) {
+	if type1 == nil || type2 == nil {
+		return false, "nil"
+	}
+	expectedType := "Unresolved"
+
+	switch type1.GetType().(type) {
+	case *sysl.Type_Primitive_:
+		expectedType = type1.GetPrimitive().String()
+		if _, ok := type2.GetType().(*sysl.Type_Primitive_); ok {
+			return type1.GetPrimitive().String() == type2.GetPrimitive().String(), expectedType
+		}
+
+	case *sysl.Type_TypeRef:
+		switch {
+		case type2.GetTypeRef() != nil:
+			ref1 := type1.GetTypeRef().GetRef()
+			ref2 := type2.GetTypeRef().GetRef()
+
+			if ref1.GetAppname() != nil && ref2.GetAppname() != nil {
+				expectedType = ref1.GetAppname().GetPart()[0]
+				return ref1.GetAppname().GetPart()[0] == ref2.GetAppname().GetPart()[0], expectedType
+			} else if ref1.GetPath() != nil && ref2.GetPath() != nil {
+				expectedType = ref1.GetPath()[0]
+				return ref1.GetPath()[0] == ref2.GetPath()[0], expectedType
+			}
+		case type1.GetTypeRef().GetRef().GetAppname() != nil:
+			expectedType = type1.GetTypeRef().GetRef().GetAppname().GetPart()[0]
+		case type1.GetTypeRef().GetRef().GetPath() != nil:
+			expectedType = type1.GetTypeRef().GetRef().GetPath()[0]
+		}
+	}
+
+	return false, expectedType
+}
+
+func resolveVariableType(expr *sysl.Expr, viewName string) (*sysl.Type, []ValidationMsg) {
+	validationMsgs := make([]ValidationMsg, 0, 10)
+
+	switch expr.Expr.(type) {
```

Comment:
```go
switch t := expr.Expr.(type) {
     case *sysl.Expr_Transform_:
               if typeRef := t.Transform.....
}
```

Reply:
I'm retrieving type in expr. Therefore t will not be used.